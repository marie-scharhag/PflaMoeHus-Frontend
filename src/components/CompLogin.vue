<template>
  <div class="complogin">
    <h1 align="center">Login</h1>
    <p id="error" align="center">{{ errormessage }}</p>
    <form @submit.prevent="loginUser()">
      <div class="row">
        <div class="col1">
          <label for="email" class="left">E-Mail Adresse</label>
        </div>
        <div class="col2">
          <input
            v-model="email"
            id="email"
            type="text"
            name="email"
            size="30"
            maxlenght="50"
            class="right"
            style="border:1px solid grey"
          />
        </div>
      </div>
      <div class="row">
        <div class="col1">
          <label for="password" class="left">Passwort</label>
        </div>
        <div class="col2">
          <input
            v-model="password"
            id="pw"
            type="password"
            name="password"
            size="20"
            maxlenght="50"
            class="right"
            style="border:1px solid grey"
          />
        </div>
      </div>
      <input type="submit" name="loginUser" value="Login" >
        <div class="row">    
            Noch kein Kunde? <router-link id="link" to="/register">Hier registrieren</router-link>
        </div>
        <div class="row">    
            Passwort vergessen? <a id="link" @click="toResetPw()" >Hier zur√ºcksetzen</a>
        </div>
    </form>
  </div>
</template>

<script lang="ts">
import { postLoginUser, useUserStore } from "../service/UserStore";
import {
  ref,
  defineComponent,
  computed,
  onMounted,
} from "vue";
import { useRouter, useRoute } from "vue-router";
import {routerHistory} from "../service/RouterStore";

export default defineComponent({
  name: "CompLogin",

  setup(props, context) {
    /**
     * users entered email
     */
    const email = ref("");
    /**
     * users entered password
     */
    const password = ref("");
    /**
     * loginrequest containing email and password
     */
    const loginRequest: LoginRequest = {
      email: email.value,
      password: password.value,
    };
    const { sendLogin, errormessage} = postLoginUser();
    const { reseterrormessage } = useUserStore();
    const router = useRouter();
    const route = useRoute();
    const COLORS = ["red", "#ccc"];

    onMounted(async () => {
      reseterrormessage();
    });

  /**
   * sends a request to the server, requesting to log in the user
   */
    async function loginUser() {
      loginRequest.email = email.value;
      loginRequest.password = password.value;
      const loginSuccess = await sendLogin(loginRequest);
      const path = route.fullPath;

      if(loginSuccess){
        if(routerHistory.getPathReversed(2) === "/register"){
          if(routerHistory.getPathReversed !== undefined){
            routerHistory.getPathReversed(4) == "/cart" ? router.push("/orderform") : router.push("/profile");
          }else{
            router.push("/");
          }
          
        }else{
          if(routerHistory.getPathReversed(2) === "/login"){
            router.push("/");
          }else{
            routerHistory.getPathReversed(2) === "/cart" ? router.push("/orderform") : router.push(routerHistory.getPathReversed(2));
          }

        }
      }
    }
    
    /**
     * changes the current component to resetting the password
     */
    function toResetPw() {
        context.emit("toggle-comp", email.value);
    }

    return {
      loginUser,
      loginRequest,
      email,
      password,
      errormessage,
      colorEmail: computed(() => {
        if (errormessage.value != "") {
          return COLORS[0];
        } else {
          return COLORS[1];
        }
      }),
      toResetPw
    };
  },
});
</script>

<style scoped lang="scss">
form {
  margin: 5% 0% 15% 35%;
}
label {
  padding: 12px 12px 12px 0;
  display: inline-block;
}
.col1 {
  float: left;
  width: 25%;
}

.col2 {
  float: left;
  width: 75%;
}

.row:after {
  content: "";
  display: table;
  clear: both;
}
input:focus{
    outline-color: $color-green;
}
input[type=text]{
    width: 30%;
    padding: 0.25em;
    border: 1px solid $color-grey3;
    border-radius:3px;
    resize: vertical;
}
input[type=submit]{
    margin: 5% 0% 5% 15%;
    padding: 1% 5%;
}

#pw {
  width: 30%;
  padding: 0.25em;
  border: 1px solid v-bind("colorEmail");
  border-radius: 3px;
  resize: vertical;
}

#email {
  width: 30%;
  padding: 0.25em;
  border: 1px solid v-bind("colorEmail");
  border-radius: 3px;
  resize: vertical;
}

#link {
    text-decoration: none;
    list-style: none;
    list-style-type: none;
    color: $color-green;
    &:hover{
        cursor: pointer;
    } 
}
</style>