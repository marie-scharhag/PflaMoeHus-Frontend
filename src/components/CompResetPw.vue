<template>
    <div class="compresetpw">
        <h1 align="center">Passwort zurücksetzen</h1>
        <p id="error" align="center">{{ errormessage }}</p>
        <p align="center">{{ message }}</p>
        <form @submit.prevent="reset()">
            <div class="row">    
                <a @click="back()" id="link" v-if="!isHidden"> &laquo; zurück</a>
            </div>
            <div class="row" v-if="!isHidden">
                <div class="col1">
                    <label for="email" class="left">E-Mail Adresse</label>
                </div>
                <div class="col2">
                    <input
                        v-model="email"
                        id="email"
                        type="text"
                        name="email"
                        size="30"
                        maxlenght="50"
                        class="right"
                    />
                </div>
            </div>
            <div v-if="!isHidden">
                <input type="submit" name="resetPw" value="Passwort zurücksetzen" >
            </div>
            <div v-if="isHidden">
                <button @click="toLogin()" > Zurück zum Login</button>
            </div>
        </form>
    </div>
</template>

<script lang = "ts">
import { defineComponent, computed, ref } from "vue";
import { useUserStore } from "../service/UserStore"
import { useEmailStore } from "../service/EmailStore"
import "@/service/Product";

export default defineComponent({
  name: "CompProducts",
  components: {},
  props: {
      mailadress: String,
  },
  setup(props, context) {
    /**
     * colors used 
     */
    const COLORS= ["red", "#ccc"];
    /**
     * mailadress entered into the form
     */
    const email = ref(props.mailadress);
    /**
     * whether the user with the given email exists
     */
    const userExists = ref(true);
    /**
     * whether an error occured during the reset of the password
     */
    const error = ref(false);
    /**
     * the message displayed when an error occured
     */
    const errormessage = ref("");
    /**
     * 
     */
    const message = ref("");

    const { checkIfEmailExists } = useUserStore();
    const { sendEmail } = useEmailStore();
    /**
     * whether a component is hidden or not
     */
    const isHidden = ref(false);

    /**
     * callback function
     */
    function back(): void {
      context.emit("toggle-comp");
    }

    /**
     * returns to the login page
     */
    function toLogin() {
        back();
    }
    /**
     * resets the page
     */
    async function reset() {

        isHidden.value = false;

        if (email.value) {
            const exists = await checkIfEmailExists(email.value);
            userExists.value = exists;
            if (!exists) {
                errormessage.value = "Diese E-Mail Adresse ist nicht bei uns Registriert.";
                message.value = "";
            } else {
                errormessage.value = "";
                message.value = "Ihnen wurde eine Email zum Passwort zurücksetzen gesendet.";
                isHidden.value = true;
                
                sendEmail(email.value);
            }
        }
    }
    

    return {
        back,
        colorEmail: computed(() => {
            if (userExists.value) {
                return COLORS[1];
            } else {
                return COLORS[0];
            }
            
        }),
        email,
        reset,
        errormessage,
        message,
        toLogin,
        isHidden
    };
  },
});
</script>

<style scoped lang="scss">
form {
  margin: 5% 0% 15% 35%;
}
label {
  padding: 12px 12px 12px 0;
  display: inline-block;
}
.col1 {
  float: left;
  width: 25%;
}

.col2 {
  float: left;
  width: 75%;
}

.row:after {
  content: "";
  display: table;
  clear: both;
}
input:focus{
    outline-color: $color-green;
}
input[type=text]{
    width: 30%;
    padding: 0.25em;
    border: 1px solid $color-grey3;
    border-radius:3px;
    resize: vertical;
}
input[type=submit]{
    margin: 5% 0% 5% 15%;
    padding: 1% 5%;
}

#email {
  width: 30%;
  padding: 0.25em;
  border: 1px solid v-bind("colorEmail");
  border-radius: 3px;
  resize: vertical;
}

#link {
    text-decoration: none;
    list-style: none;
    list-style-type: none;
    color: $color-green;
    &:hover{
        cursor: pointer;
    } 
}

#error {
    color: red;
}
</style>