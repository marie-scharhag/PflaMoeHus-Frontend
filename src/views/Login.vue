<template>
  <div class="login">
    <component v-bind:is="compref" @toggle-comp="toggle($event)" :mailadress="email"/>
  </div>
</template>

<script lang="ts">
import CompLogin from "../components/CompLogin.vue"
import CompResetPw from "../components/CompResetPw.vue"
import { defineComponent, ref, onMounted } from "vue"
import '../service/Product'
import {routerHistory} from "../service/RouterStore";

export default defineComponent({
    name: "Login",
    components:{
        CompLogin,
        CompResetPw
    },
    setup() {
        /**
         * options of visible components
         */
        const COMPONENTS = ["CompLogin", "CompResetPw"];
        /**
         * default component is login
         */
        const component = COMPONENTS[0];
        /**
         * currently visible component
         */
        const compref = ref(component);

        /**
         * a user's email adress
         */
        const email = ref("");

        routerHistory.add("/login");

        /**
         * changes the currently visible component between login and password reset
         * @param e a user's entered email adress
         */
        function toggle(e: string): void {
            email.value = e;
            if (compref.value === COMPONENTS[0]) {
                compref.value = COMPONENTS[1];
            } else {
                compref.value = COMPONENTS[0];
            }
        }

        return {
            component,
            compref,
            toggle,
            email
        };
    }
});
</script>