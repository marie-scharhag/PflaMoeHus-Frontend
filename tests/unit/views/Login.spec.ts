import Login from "@/views/Login.vue"
import CompResetPw from "@/views/Login.vue"
import CompLogin from "@/views/Login.vue"

import { shallowMount } from "@vue/test-utils"

describe('Login.vue',() => {

    const cmp = shallowMount(Login)

    test('test html and structur with snapshots',() => {
        expect(cmp.html()).toMatchSnapshot()
        expect(cmp).toMatchSnapshot()
    })

    test('Find component CompResetPw CompLogin and test if they exist', () => {

        expect(cmp.getComponent(CompResetPw)).toBeTruthy()
        expect(cmp.getComponent(CompLogin)).toBeTruthy()

        expect(() => cmp.getComponent('.not-there')).toThrowError()
    })
})