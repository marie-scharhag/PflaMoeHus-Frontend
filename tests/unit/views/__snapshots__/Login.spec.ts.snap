// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue test html and structur with snapshots 1`] = `
<div class="login">
  <comp-login-stub mailadress=""></comp-login-stub>
</div>
`;

exports[`Login.vue test html and structur with snapshots 2`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {},
      "__props": Array [
        Object {},
        Array [],
      ],
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
    >
      <div
        class="login"
      >
        <comp-login-stub
          mailadress=""
        />
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {
        "transition": Object {
          "name": "transition",
          "props": Object {},
          "render": [Function],
        },
        "transition-group": Object {
          "name": "transition-group",
          "props": Object {},
          "render": [Function],
        },
      },
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
        Object {
          "__emits": null,
          "__props": Array [],
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.5",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "__emitted": Object {},
    "component": "CompLogin",
    "compref": "CompLogin",
    "email": "",
    "toggle": [Function],
  },
  "rootVM": Object {
    "__emitted": Object {},
  },
}
`;
